client/starwhale/api/_impl/dataset/loader.py

```diff
@@ -203,23 +265,32 @@ def _iter_meta(self) -> t.Generator[TabularDatasetRow, None, None]:
                             lk.signed_uri = uri_dict.get(lk.uri, "")
 
                         for row in rows:
+                            rows_cnt += 1
                             yield row
                 else:
                     for row in self.tabular_dataset.scan(rt[0], rt[1]):
+                        rows_cnt += 1
                         yield row
 
-    def _iter_meta_with_queue(self, mq: queue.Queue[_TMetaQItem]) -> None:
-        # TODO: tune last processed range
+                self._key_range_dict[(rt[0], rt[1])] = {
+                    "rows_cnt": rows_cnt,
+                    "processed_cnt": 0,
+                }
+
+    def _iter_meta_for_queue(self) -> None:
+        out_mq = self._meta_fetched_queue
+        if out_mq is None:
+            raise RuntimeError("queue not initialized for iter meta")
         try:
             for meta in self._iter_meta():
                 if meta and isinstance(meta, TabularDatasetRow):
-                    mq.put(meta)
+                    out_mq.put(meta)
         except Exception as e:
-            mq.put(e)
+            out_mq.put(e)
             raise
 
         for _ in range(0, self._num_workers):
```

Comment:
use out_mq.task_done() to replace foreach? the same with function of '_unpack_row_for_queue'